### ISSUE TITLE
[RFC] Use CUDA graphs by default on torch.compile

### ISSUE THREAD

**Issue by lezcano** [pytorch/pytorch#121968]
### 🐛 Describe the bug

I keep seeing comparisons between JAX / TF / Keras vs `torch.compile` where they benchmark the "default XLA settings vs. the default `torch.compile`" to find that XLA frontends are x3-x4 faster than `torch.compile`.
The last one I found is the newly posted Keras 3 benchmarks
https://keras.io/getting_started/benchmarks/ 
but I have seen these skewed comparisons over and over again.

The only thing stopping us from compiling with CUDA graphs on is that these sometimes have a higher mem footprint. Would it be reasonable to turn these on by default so that people get speed by default, and then catch a potential OOM and re-raise it indicating that using CUDA-graphs off to potentialy mitigate the OOM?

cc @ezyang @jansel @albanD @eellison  @Chillee 

### Versions

master

**Comment by bsergean** [comment]
Q: can AOT Inductor ([this](https://github.com/pytorch/pytorch/blob/main/docs/source/torch.compiler_aot_inductor.rst)) works with cuda graphs ? We did some testing a while back and it wouldn't work, but we would for sure love those 3x/4x speedup you are talking about !

**Comment by lezcano** [comment]
@desertfire for the AOT Inductor question

**Comment by eellison** [comment]
We could potentially disable cudagraphs on OOM, and hook back the cudagraph tensors into the cudacaching allocator (similar to what we do now with cudagraph checkpointing).

**Comment by jansel** [comment]
We could also detect a likely-overhead-bound model (small shapes) and issue a warning to enable cudagraphs.

**Comment by Skylion007** [comment]
I've had some real world workloads like trying to compile Caduceus: https://github.com/kuleshov-group/caduceus that caused some memory leaks with torch CUDAGraphs (with max-autotune). I suspect enabling it by default will raise a lot more of these issues in practice, just an FYI.

The memory leak took a while to fill up the GPU memory so running it once and catching the OOM may not work in this scenario.

**Comment by peterbell10** [comment]
@Skylion007 is there an issue open for this memory leak?

**Comment by ezyang** [comment]
CUDA graphs has operational implications. Maybe we can enable it by default in very limited situations where we know it definitely will not have problems, but those will have to be carefully analyzed.

**Comment by lezcano** [comment]
@ezyang can you elaborate on these "operational implications"?

**Comment by ezyang** [comment]
For example, if you have dynamic shapes, then you need to make sure you don't have *too* many dynamic shapes, because if you do we will keep cuda graphing each of them and you'll OOM. Cuda graphs also makes some things expensive (must copy inputs into cuda graph memory) and our strategy for dealing with graph breaks in cuda graphs means some operations are illegal (e.g., keeping references to the cuda graph memory region live while you want to run that cuda graph again).

**Comment by ifsheldon** [comment]
I don't know if it's helpful for this issue, but just FYI, CUDA 12.4 introduces conditional nodes and loop nodes, like IF and WHILE. 

In theory, all TensorFlow 1.x models can be converted into massive CUDA graphs. But I don't know what are the technical difficulties in the case of PyTorch and `torch.compile` as I am not very familiar with the technical details.

**Comment by lezcano** [comment]
>For example, if you have dynamic shapes, then you need to make sure you don't have too many dynamic shapes, because if you do we will keep cuda graphing each of them and you'll OOM.

What about turning them on when `dynamic=False` or there are no dynamic shapes? Those are the cases that XLA supports and they always have them on then.

**Comment by ezyang** [comment]
Ok sure but there are other gotchas

**Comment by Chillee** [comment]
@ezyang I think it's a question whether those gotchas are worth the benefits of turning on cudagraphs by default. Since not turning on cudagraphs by default is also arguably a "gotcha".

**Comment by ezyang** [comment]
I will feel happier about turning it on by default if we put in some effort to define policy for when it is a good idea or not. I will be unhappy if we just YOLO turn it on by default.

**Comment by galv** [comment]
> I don't know if it's helpful for this issue, but just FYI, CUDA 12.4 introduces conditional nodes and loop nodes, like IF and WHILE, which is epic IMHO.
> 
> In theory, all TensorFlow 1.x models can be converted into massive CUDA graphs. But I don't know what are the technical difficulties in the case of PyTorch and `torch.compile` as I am not very familiar with the technical details.

So I've been looking into this from the NVIDIA side. I actually have an intern working on getting some. I don't think that IF and WHILE nodes will necessarily solve the problems discussed here. They can reduce the number of graph breaks in torch dynamo, by doing dynamic control flow on the GPUs rather than on the CPU, for torch.cond and torch.while_loop. I have an intern looking at this for some algorithms in NeMo right now: https://github.com/NVIDIA/NeMo/pull/9212/files#diff-acab1a9f3d702862ddbe5720bfa6c7fd0a57f7c3dc0b59eb9878ed5cd1e3513aR61-R69

The aforementioned problem about having to copy inputs can actually be overcome as of CUDA 12.4 using [cudaGraphKernelNodeUpdatesApply](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH_1g2d558cf37c9616365c67447e61ac0d6a)(). I began making a prototype, but deprioritized it for the work on seeing if we can get torch.cond and torch.while_loop to use conditional nodes, rather than suffer performance penalties from graph breaks.

The problem of handling dynamic shapes is another matter, because ideally memory would be shared among every cuda graph specialized for a particular input shape (assuming that only one cuda graph runs at a time) for a function of interest. This is a little tricky, but I think we can make it happen with the low level virtual memory mapping APIs in the cuda driver API.

Feel free to contact me if anyone is interested in these topics.

**Comment by lezcano** [comment]
Another example of an important repo benchmarking against `torch.compile` without using CUDA graphs:
https://github.com/karpathy/build-nanogpt/issues/6

By looking at the benchmarks, comparing `default` and `cudagraph_dynamic`, it looks like in many reasonable scenarios, having `reduce-overhead` on by default is exactly what the user wants.

![image](https://github.com/pytorch/pytorch/assets/3291265/465dec66-915b-4390-bc73-922b212cb3e9)
Now, if big players like Karpathy or Keras core devs miss turning on this feature, I think it is very likely that most of people using `torch.compile` are also missing this feature when running their models.

Here, I would argue that's better to ask for forgiveness than fro premission, and have CUDA graphs on by default. Then, we could modify the error we throw when OOMing telling the user that they can try compiling their graph without CUDA graphs.

**Comment by eellison** [comment]
@fxmarty we did something with warning after a certain number of shapes recently. we could have an option to toggle the warning into a fallback. note, there can be some memory issues with duplicating eager and cuda graph pool. cc @BoyuanFeng https://github.com/pytorch/pytorch/pull/131387

**Comment by fxmarty** [comment]
from: https://github.com/pytorch/pytorch/issues/128424#issuecomment-2161919814

> Or you can rearchitect your prefill/decode so that it is CUDA graph friendly, as was done in gpt-fast.

I think that is what we need to do cc @gante

**Comment by ezyang** [comment]
> For example, say I would like to have cuda graphs captured only for shapes [batch_size, 1], [batch_size, 2], ..., [batch_size, 16] but not above that?

The big question I'd ask is, what do you want to happen if you don't have a power of two shape? If you propose to run without cuda graphs, well, you've traded one perf footgun for another :P

**Comment by fxmarty** [comment]
@ezyang I meant having a way to specify which shapes we would like to use cuda graphs for (e.g. 1, 2, 3, 4, ..., 14, 15, 16).

> If you propose to run without cuda graphs, well, you've traded one perf footgun for another :P

Well, this is what is done as well in https://pytorch.org/blog/accelerating-generative-ai-2/ tbh. The footgun of cuda graph capture is much bigger than the footgun of not using cuda graph at all, especially in compute bound regime.

**Comment by ezyang** [comment]
OK, seems reasonable to ask for

**Comment by vadimkantorov** [comment]
Do CUDA graphs specialize in the input tensor shape (and stride?) of the underlying storage size?

E.g. can we send in a flat tensor with padding by powers of two and inside the function/graph just take a slice (thus removing the padding) and then reshaping this chunk to the original shape to restore the logic?

This could enable CUDA graphs for NJT tensors?

**Comment by BoyuanFeng** [comment]
> Do CUDA graphs specialize in the input tensor shape (and stride?) of the underlying storage size?
> 
> E.g. can we send in a flat tensor with padding by powers of two and inside the function/graph just take a slice (thus removing the padding) and then reshaping this chunk to the original dynamic shape to restore the logic?
> 
> This could enable CUDA graphs for NJT tensors? (via preallocating a large-sized input flat tensor or having a collection of preallocates large-sized input flat tensors)

CUDAGraph specializes on the input tensor shape/strides. We cannot pass a padded tensor, slice, and then operate on the sliced tensors.

**Comment by ezyang** [comment]
That's precisely it. Traditionally, the slice and then compute is handled host side (with strides) which is why it is not Cuda graphable. But if you rewrite the kernels so that the slice management happens in the kernels (at some runtime cost!) then Cuda graphs is possible again. I do not think this is a bad idea, especially when you have a compiler to do the work for you. But it is Work(TM)

**Comment by vadimkantorov** [comment]
@ezyang could it even by some small additional preproc kernel which does recompute strides if needed? Also, if it's simply a padded tensor, maybe even no stride recomputes are needed? probably even now kernels bail-out processing outside of the numel? so that should already auto-magically handle padding on the right? no?

Also a question: so does it mean that NJT is not supported with CUDA Graphs currently?

**Comment by ezyang** [comment]
Unfortunately, the kernels themselves need to change, because the strides are typically sent as part of the launch arguments, but you need them to be computed from GPU memory. Padding though I agree should work.

And yes, I would be extremely surprised if NJT had worked with CUDA graphs

**Comment by vadimkantorov** [comment]
I guess yes, for this kernel needs to load the strides/effective numel from GPU memory. But it seems that the upside of much more flexible CUDA graph support should also be huge? Is the problem that loading these from the GPU memory would be much slower than from the kernel arguments memory? Because otherwise one could imagine passing some sorts of tensor descriptor structs in gpu memory. Or having some sort of two alternative kernel launchers (both of which call the same underlying device kernel code)? E.g. do CUDNN kernels support reading descriptors from GPU memory?

And given that NJT likely does not have strides per se, the offsets etc are already stored in GPU memory, so it should work?

**Comment by ezyang** [comment]
Well the big problem is typically you will run a different kernel entirely depending on if the input is contiguous or not lol. In any case, it's still possible, especially if you do things like "interpreter on the GPU" shenanigans, it's just work!

**Comment by vadimkantorov** [comment]
@ezyang I'm also wondering, are nested tensors still relevant?

As it appears, in torchtune to benefit from the features of regular dense tensors (and their composability with distributed features etc), they had to reinvent these as "padded packed tensors":
- https://github.com/pytorch/torchtune/pull/2819#issuecomment-2977047801

If not, maybe worth exposing more raw version of sdpa kernels which accepts varlen dense tensors along with seqlens, in order to complement the NJT subclass version (especially if NJT development is paused for now)

**Comment by ezyang** [comment]
Yeah, I'm with you on that. @drisspg @jbschlosser thoughts?

**Comment by drisspg** [comment]
ill let Joel speak to the NJT comment but you are not the first to suggest
> exposing more raw version of sdpa kernels which accepts varlen dense tensors along with seqlens

and yeah plan to expose this public API, still working on an RFC doc for API, ill tag you when ready

**Comment by vadimkantorov** [comment]
Because otherwise, people still have to depend on flash_attn / xformers to access these kernels. And this varlen version becomes more mainstream...

**Comment by galv** [comment]
> We cannot pass a padded tensor, slice, and then operate on the sliced tensors.

IMO, we could do it via SWITCH conditional statements, but it is going to be a lot of work and not worth waiting for since there are some technical risks.
